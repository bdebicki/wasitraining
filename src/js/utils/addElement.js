export default function(tag, {
	clsssNames,
	id,
	data,
	children,
}) {
	const element = document.createElement(tag);
	const setData = (dataAttr) => {
		const dataName = Object.keys(dataAttr)[0];
		element.dataset[dataName] = dataAttr[dataName];
	};

	if (typeof (clsssNames) === 'string') {
		element.classList.add(clsssNames);
	} else if (Array.isArray(clsssNames)) {
		clsssNames.forEach((className) => element.classList.add(className));
	}

	if (id) {
		element.id = id;
	}

	if (typeof (data) === 'object' && !Array.isArray(data)) {
		setData(data);
	} else if (Array.isArray(data)) {
		data.forEach((dataAttr) => {
			setData(dataAttr);
		});
	}

	if (typeof (children) === 'string') {
		element.textContent = children;
	} else if (typeof (children) === 'object') {
		element.appendChild(children);
	}

	return element;
}
