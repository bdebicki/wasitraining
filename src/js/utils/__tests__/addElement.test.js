import addElement from '../addElement';
import cleanDOM from '../../../../tests/utils/cleanDOM';
import using from '../../../../tests/utils/using';
import spyConsole from '../../../../tests/utils/spyConsole';

const elementTag = 'span';
const addElementCases = [
	{
		description: 'simple <span> element',
		tag: elementTag,
	},
	{
		description: '<span> element with id and title',
		tag: elementTag,
		settings: {
			id: 'spanId',
			title: 'span element tile',
		},
	},
	{
		description: 'link <a> element with href attribute',
		tag: 'a',
		settings: {
			href: 'http://google.com',
		},
	},
	{
		description: 'element with single class name',
		tag: elementTag,
		settings: {
			classNames: 'spanEl',
		},
	},
	{
		description: 'element with single data attribute',
		tag: elementTag,
		settings: {
			dataAttr: { spanElDataAttr: 'data value' },
		},
	},
	{
		description: 'element with multiple data attribute',
		tag: elementTag,
		settings: {
			dataAttr: [
				{ firstDataAttr: 'first data attr value' },
				{ secondDataAttr: 'second data attr value' },
			],
		},
	},
	{
		description: 'element with text',
		tag: elementTag,
		settings: {
			children: 'span element',
		},
	},
];

afterAll(() => cleanDOM());

describe('addElement function tests', () => {
	spyConsole();

	using(addElementCases).describe('render html element using addElement', ({ description, tag, settings }) => {
		it(description, () => {
			const el = addElement(tag, settings);

			expect(el).toMatchSnapshot();
		});
	});
	it('test element with click event listener', () => {
		// having
		const el = addElement(elementTag, {
			classNames: 'spanEl',
			onClick: () => console.log(),
		});
		document.body.appendChild(el);

		// when
		document.querySelector('.spanEl').click();

		// then
		expect(console.log).toHaveBeenCalled();
	});
});
