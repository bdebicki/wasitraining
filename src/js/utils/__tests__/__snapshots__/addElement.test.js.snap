// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addElement function tests render html element using addElement <span> element with id and title 1`] = `
<span
  id="spanId"
  title="span element tile"
/>
`;

exports[`addElement function tests render html element using addElement element with multiple data attribute 1`] = `
<span
  data-first-data-attr="first data attr value"
  data-second-data-attr="second data attr value"
/>
`;

exports[`addElement function tests render html element using addElement element with single class name 1`] = `
<span
  class="spanEl"
/>
`;

exports[`addElement function tests render html element using addElement element with single data attribute 1`] = `
<span
  data-span-el-data-attr="data value"
/>
`;

exports[`addElement function tests render html element using addElement element with text 1`] = `
<span>
  span element
</span>
`;

exports[`addElement function tests render html element using addElement link <a> element with href attribute 1`] = `
<a
  href="http://google.com"
/>
`;

exports[`addElement function tests render html element using addElement simple <span> element 1`] = `<span />`;
